compiler/rustc_resolve/src/check_unused.rs

```diff
@@ -327,7 +327,45 @@ impl Resolver<'_> {
                         || module.ends_with("_test")
                         || module.ends_with("_tests") =>
                 {
-                    Some(parent_module.span)
+                    // If the parent module has the `#[cfg(test)]` attribute, suppress a help to add
+                    // the attribute; however, if not, suggest adding the attribute. There is no way
+                    // to retrieve attributes here because we no longer have `TyCtxt`. Thus, we must
+                    // detect the attribute by manipulating the source code as a string to avoid
+                    // redundant suggestions.
+                    let sm = visitor.r.session.source_map();
+                    match sm.span_extend_to_prev_str(
```

Comment:
The source_map should only be used as a last resort when building diagnostics.
The code path below can only happen if we are reporting a diagnostic while in `--cfg test` mode, isn't it?
Can't we just check for it?

Reply:
@cjgillot 
Thank you for your review.

I think the case that a mod does not have the `#[cfg(test)]` attribute should show both warning and help.

```rust
mod tests {
    use std::collections::HashSet;

    #[test]
    fn test_empty() {}
}

fn main() {}
```

:o:
```bash
$ rustc test.rs
warning: unused import: ...
...
help: consider adding a `#[cfg(test)]` ...
```

:o:
```bash
$ rustc --test test.rs
warning: unused import: ...
...
help: consider adding a `#[cfg(test)]` ...
```

However, when we have a mod that has the attribute, then we should avoid emitting the help even if we are in the `--cfg test` mode.

```rust
#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    #[test]
    fn test_empty() {}
}

fn main() {}
```

:o:
```bash
$ rustc test.rs
# nothing
```

:x:
```bash
$ rustc --test test.rs
warning: unused import: ...
...
# ↓ incorrect: we want to suppress this help.
help: consider adding a `#[cfg(test)]` ...
```

So, if my understanding is correct, we still need to check if a mod has the attribute regardless of the `--cfg test` mode.

Could you please tell me if there is a way to implement this without using `source_map`?

Reply:
Can we get `rustc_ast::ast::Item` from `parent_module`?

If so, we may retrieve the `#[cfg(test)]` attribute from [`attrs`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_ast/ast/struct.Item.html#structfield.attrs), though.

Reply:
You don't really need to check for the presence of the attribute, you just need to know whether we are in test mode.
I see 4 cases:

1.  no cfg(test) and no --test: there is unused-import to report, so print both the warning and help;
2.  cfg(test) and no --test: there is no unused-import to report, so nothing to do;
3.  no cfg(test) and --test: there is unused-import to report, so print the warning and no help;
4.  cfg(test) and --test: there is unused-import to report, so print the warning and no help.

In case 2, the loop for unused in visitor.unused_imports is empty, so there is nothing to do.
To distinguish between case 1 and 3-4, you just need to check for --test mode to know if you need to print the help.
You don't need to check the attributes at any time.

Reply:
@cjgillot 
Oh, I see. Thank you :)